04.ls/ls.rb

```diff
@@ -31,6 +34,70 @@ def display_filename_matrix(file, column)
   end
 end
 
+# ここからlオプション
+PERMISSION_CODES = {
+  '7' => 'rwx',
+  '6' => 'rw-',
+  '5' => 'r-x',
+  '4' => 'r--',
+  '3' => '-wx',
+  '2' => '-w-',
+  '1' => '--x',
+  '0' => '---'
+}.freeze
+
+def permission_and_filetype(file_detail_information)
+  filetype = file_detail_information.ftype == 'file' ? '-' : 'd'
+  permission = file_detail_information.mode.to_s(8).slice(-3..-1).chars
+  filetype + permission.map { |n| PERMISSION_CODES[n] }.join
+end
+
+def last_updated(file_detail_information)
+  last_updated = file_detail_information.mtime
+  half_year_ago = Date.today << 6
+  time_or_year = last_updated.to_date >= half_year_ago ? last_updated.strftime('%H:%M') : last_updated.year.to_s.rjust(5)
+  "#{last_updated.mon.to_s.rjust(2)} #{last_updated.day.to_s.rjust(2)} #{time_or_year}"
+end
+
+def collect_file_details(file_detail_information)
+  file_detail_information.map do |file|
+    permission_and_filetype = permission_and_filetype(File::Stat.new(file))
+    last_updated = last_updated(File::Stat.new(file))
+    {
+      permission: permission_and_filetype,
+      nlink: File::Stat.new(file).nlink.to_s,
+      username: Etc.getpwuid(File::Stat.new(file).uid).name,
+      groupname: Etc.getgrgid(File::Stat.new(file).gid).name,
+      filesize: File::Stat.new(file).size.to_s,
+      time: last_updated,
+      filename: file
+    }
```

Comment:
一旦変数にいれる必要がないように思いました。
```suggestion
    {
      permission: permission_and_filetype(File::Stat.new(file)),
      nlink: File::Stat.new(file).nlink.to_s,
      username: Etc.getpwuid(File::Stat.new(file).uid).name,
      groupname: Etc.getgrgid(File::Stat.new(file).gid).name,
      filesize: File::Stat.new(file).size.to_s,
      time: last_updated(File::Stat.new(file)),
      filename: file
    }
```